File: backend/app/services/checklist.py
"""Checklist forms: a table of requirements, each answered Yes, No or N/A.

Most inspection forms are drawn on a canvas and read as a list of questions.
A checklist is different: it is a regulator's table - indicators, each with its
numbered compliance requirements - and the inspection passes only when every
requirement is met. So a checklist is stored as data, not as a layout, and it
decides for itself whether the inspection can pass.

Shape, kept under its own key so the canvas builder never mistakes it for one
of its own layouts:

    {"title": ..., "version": 1, "source": "phealth_checklist",
     "checklist": {"sections": [
        {"code": "Ind 01", "title": "...", "requirements": [
            {"code": "1.1", "text": "..."},
            {"code": "2.1", "text": "...", "alternative": "access"},
            {"code": "2.2", "text": "...", "alternative": "access"}]}]}}

Requirements sharing an ``alternative`` are either-or ("2.1 ... OR 2.2 ..."):
the pair is met when any one of them is met, so answering No to the one that
does not apply cannot fail an inspection that is otherwise fine.
"""
from __future__ import annotations

from typing import Any, Iterable, Optional

SOURCE = "phealth_checklist"
ANSWERS = ("yes", "no", "na")


def is_checklist(schema: Any) -> bool:
    return isinstance(schema, dict) and schema.get("source") == SOURCE and isinstance(schema.get("checklist"), dict)


def requirements(schema: dict) -> list[dict]:
    rows = []
    for section in schema.get("checklist", {}).get("sections", []) or []:
        for requirement in section.get("requirements", []) or []:
            rows.append({**requirement, "section": section.get("code")})
    return rows


def evaluate(schema: dict, answers: Optional[dict]) -> dict[str, Any]:
    """Where a filled checklist stands, and whether it can pass.

    It passes when every requirement is answered and none is unmet. An
    either-or group counts once: met if any member is Yes, not applicable if
    every member is N/A, unmet otherwise.
    """
    answers = {str(key): str(value).lower() for key, value in (answers or {}).items()}
    rows = requirements(schema)
    unanswered = [row["code"] for row in rows if answers.get(row["code"]) not in ANSWERS]

    groups: dict[str, list[dict]] = {}
    singles: list[dict] = []
    for row in rows:
        (groups.setdefault(row["alternative"], []) if row.get("alternative") else singles).append(row)

    met, not_met, not_applicable = [], [], []
    for row in singles:
        value = answers.get(row["code"])
        if value == "yes":
            met.append(row["code"])
        elif value == "no":
            not_met.append(row["code"])
        elif value == "na":
            not_applicable.append(row["code"])
    for members in groups.values():
        values = [answers.get(row["code"]) for row in members]
        codes = [row["code"] for row in members]
        if "yes" in values:
            met.extend(codes)
        elif all(value == "na" for value in values):
            not_applicable.extend(codes)
        elif all(value in ANSWERS for value in values):
            not_met.append(" or ".join(codes))

    return {
        "total": len(rows),
        "answered": len(rows) - len(unanswered),
        "met": len(met),
        "not_met": not_met,
        "not_applicable": len(not_applicable),
        "unanswered": unanswered,
        "can_pass": not unanswered and not not_met,
    }


def summary_line(result: dict) -> str:
    if result["unanswered"]:
        return "{} of {} requirements still to answer".format(len(result["unanswered"]), result["total"])
    if result["not_met"]:
        return "{} requirement{} not met: {}".format(
            len(result["not_met"]), "" if len(result["not_met"]) == 1 else "s", ", ".join(result["not_met"]))
    return "All requirements met"


def check_answers(forms: Iterable[tuple[int, str, dict]], submitted: Optional[list]) -> list[dict]:
    """Evaluate every checklist among the item's forms against what was sent.

    ``forms`` is (form_id, name, schema); ``submitted`` is the list the screen
    sends, one entry per form with its answers.
    """
    by_form = {int(entry.get("form_id")): entry.get("answers") or {}
               for entry in (submitted or []) if isinstance(entry, dict) and entry.get("form_id") is not None}
    results = []
    for form_id, name, schema in forms:
        if not is_checklist(schema):
            continue
        result = evaluate(schema, by_form.get(form_id))
        results.append({"form_id": form_id, "name": name, **result})
    return results

File: frontend/src/pages/Inspections/programme/checklist.ts
/**
 * Checklist forms, as the inspection panel reads them.
 *
 * Mirrors app/services/checklist.py, which is what actually decides: the
 * server refuses to record a pass the checklist does not allow. This copy
 * exists so the tally and the Passed button agree with the server while the
 * inspector is still filling it in, rather than after they press it.
 */
export type ChecklistAnswer = 'yes' | 'no' | 'na'

export interface ChecklistRequirement {
  code: string
  text: string
  /** Requirements sharing this are either-or: meeting one meets them all. */
  alternative?: string
}

export interface ChecklistSection {
  code: string
  title: string
  requirements: ChecklistRequirement[]
}

export interface ChecklistSchema {
  title?: string
  version?: number
  source: 'phealth_checklist'
  checklist: { sections: ChecklistSection[] }
}

export interface ChecklistResult {
  total: number
  answered: number
  met: number
  notMet: string[]
  notApplicable: number
  unanswered: string[]
  canPass: boolean
}

export const isChecklist = (schema: unknown): schema is ChecklistSchema => {
  const value = schema as Partial<ChecklistSchema> | null | undefined
  return Boolean(value && value.source === 'phealth_checklist' && value.checklist
    && Array.isArray(value.checklist.sections))
}

export const requirementsOf = (schema: ChecklistSchema): ChecklistRequirement[] =>
  schema.checklist.sections.flatMap((section) => section.requirements)

export function evaluateChecklist(
  schema: ChecklistSchema, answers: Record<string, string | undefined>,
): ChecklistResult {
  const rows = requirementsOf(schema)
  const valid = (value?: string): value is ChecklistAnswer => value === 'yes' || value === 'no' || value === 'na'
  const unanswered = rows.filter((row) => !valid(answers[row.code])).map((row) => row.code)

  const groups = new Map<string, ChecklistRequirement[]>()
  const singles: ChecklistRequirement[] = []
  for (const row of rows) {
    if (row.alternative) groups.set(row.alternative, [...(groups.get(row.alternative) ?? []), row])
    else singles.push(row)
  }

  let met = 0
  let notApplicable = 0
  const notMet: string[] = []
  for (const row of singles) {
    const value = answers[row.code]
    if (value === 'yes') met += 1
    else if (value === 'no') notMet.push(row.code)
    else if (value === 'na') notApplicable += 1
  }
  for (const members of groups.values()) {
    const values = members.map((row) => answers[row.code])
    if (values.includes('yes')) met += members.length
    else if (values.every((value) => value === 'na')) notApplicable += members.length
    else if (values.every(valid)) notMet.push(members.map((row) => row.code).join(' or '))
  }

  return {
    total: rows.length,
    answered: rows.length - unanswered.length,
    met,
    notMet,
    notApplicable,
    unanswered,
    canPass: unanswered.length === 0 && notMet.length === 0,
  }
}

export function checklistSummary(result: ChecklistResult): string {
  if (result.unanswered.length) {
    return `${result.unanswered.length} of ${result.total} requirements still to answer`
  }
  if (result.notMet.length) {
    return `${result.notMet.length} requirement${result.notMet.length === 1 ? '' : 's'} not met: ${result.notMet.join(', ')}`
  }
  return 'All requirements met'
}
